File: marathon-python/src/libpy/marathon/results.py
from net.sourceforge.marathon.api import SourceLine
from java.lang import System
import jarray
import sys

class collector:
	def __init__(self, result, exclude):
		self.playbackresult = result
		self.exclude = exclude

	def callprotected(self, function, result, *args):
		self.playbackresult = result
		import java.lang.Throwable
		try:			
			function.__call__(*args)
			return 1
		except java.lang.Throwable:
			(exc_type, e, tb) = sys.exc_info()
			self.addjavaerror(e, tb)
			raise
		except:
			(exc_type, e, tb) = sys.exc_info()
			self.addpythonerror(exc_type, e, tb)
			raise
		return 0

	def addfailure(self, message, result):
		self.playbackresult = result
		assertion.assertionFailed()
		frame = sys._getframe(0)
		self._addfailure(message, frame)

	def addpythonerror(self, exception_type, exception, traceback):
		try:
			self._addfailure(str(exception_type) + ': ' + exception.__str__(), _firstframe(traceback))
		except:
			self._addfailure("Unknown Error:",_firstframe(traceback))

	def addjavaerror(self, exception, traceback):
		self._addfailure(exception.getMessage(), _firstframe(traceback))

	def _addfailure(self, message, frame):
		lines = []
		while frame != None:
			code = frame.f_code
			filename = code.co_filename
			print filename
			dir = System.getProperty('marathon.project.dir')
			if filename.find(dir) == 0 or filename.find('Untitled') == 0:
				lines.append(SourceLine(code.co_filename, code.co_name, frame.f_lineno))

			frame = frame.f_back

		self.playbackresult.addFailure(message, jarray.array(lines, SourceLine))

def _firstframe(traceback):
	while traceback.tb_next != None:
		traceback = traceback.tb_next
	return traceback.tb_frame

File: marathon-ruby/src/marathon/results.rb
# $Id: results.rb 260 2009-01-13 05:53:15Z kd $
# 
java_import 'net.sourceforge.marathon.api.SourceLine'

class Collector
  def initialize()
    @playbackresult = nil
    @exclude = nil
  end

  def callprotected(function, result, *args)
    @playbackresult = result
      begin
        function.call(*args)
		  return 1
		rescue NativeException => e
		  addjavaerror(e)
		  raise
		rescue => e
		  addpythonerror(e)
		  raise
      end
      return 0
  end

  def addfailure(exception, result)
    @playbackresult = result
    $assertion.assertionFailed()
    backtrace = nil
    begin
      raise NameError
    rescue
      backtrace = $!.backtrace
    end
	_addfailure(exception.message, backtrace)
  end

  def addpythonerror(exception)
    _addfailure(exception.to_s + ": " + exception.message.to_s, exception.backtrace)
  end

  def addjavaerror(exception)
    _addfailure(exception.message, exception.backtrace)
  end

  def _addfailure(message, backtrace)
    lines = convert(backtrace)
    @playbackresult.addFailure(message, lines.to_java(SourceLine))
  end

  def convert(backtrace)
    backtrace.find_all { |item| not excluded(item) }.map { |item|
      item = item[6,item.length] if(item.index('file:/') == 0)
      matched = item.match(/(.*):(.*):(.*)/).to_a
      fname = "Unknown"
      fname = matched[3].match("`(.*)'").to_a[1] if matched[3]
      SourceLine.new(matched[1], fname, matched[2].to_i)
    }
  end

  def excluded(item)
    item = item[6,item.length] if(item.index('file:/') == 0)
    item =~ /(.*):(.*):(.*)/
    file = $1
    dir = java.lang.System.getProperty 'marathon.project.dir'
    return file.index(dir) != 0 && file.index('Untitled') != 0
  end
end
